### PR TITLE
Support only new step API (while retaining compatibility functions) 

### PR DESCRIPTION
**Pull request by arjun-kg**
# Description

This removes backward compatibility for done (old) step API (refer #2752 for details on terminated truncated (new) vs done (old) API). 
Once merged, gym will officially support only terminated truncated (new) step API. To use done (old)  API, for convenience, compatibility functions and a compatibility wrapper at make are still retained. 

The changes made for this PR includes, 
- [x] Removing `new_step_api` arguments from vector and wrapper classes, and removing all compatibility code.
- [x] Switching `new_step_api` default to `True` for compatibility functions, and `StepAPICompatibility` wrapper
- [x] Updating all existing tests to support new step API
- [x] Updating `play.py`

## Type of change

Please delete options that are not relevant.

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: gym/envs/registration.py [hunk 1]
```diff
@@ -547,7 +547,7 @@ def make(
     id: Union[str, EnvSpec],
     max_episode_steps: Optional[int] = None,
     autoreset: bool = False,
-    new_step_api: bool = False,
+    new_step_api: bool = True,
```
Comment (pseudo-rnd-thoughts) [inline gym/envs/registration.py:550]: Rename to 'apply_step_compatibility' as in a few years time new or old won't be descriptive

File: gym/vector/__init__.py [hunk 2]
```diff
@@ -15,7 +15,6 @@ def make(
     asynchronous: bool = True,
     wrappers: Optional[Union[callable, List[callable]]] = None,
     disable_env_checker: Optional[bool] = None,
-    new_step_api: bool = False,
```
Comment (pseudo-rnd-thoughts) [inline gym/vector/__init__.py:18]: Same parameter name suggestion

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Overall, looks good. I only have one suggestion on the same change. I realise that this is not backward compatible but for the future, I think that it is better. 
If there is a problem with v25 compatibility, then there may be enough changes to justify a v25.1 with this parameter change.

**Comment by pseudo-rnd-thoughts** [comment]
In #3028, I have removed the language changes such that v0.25.3 will be backward compatible.
Therefore, would you be able to make the language changes in this PR. 
These are my proposed changes

* `OldStepType` -> `DoneStepType`
* `NewStepType` -> `TerminationTruncationStepType`
* `step_to_new_api()` -> `convert_to_terminated_truncated_step_api()`
* `step_to_old_api()` -> `convert_to_done_step_api()`
* `new_step_api` -> `convert_to_termination_truncation`

File: gym/utils/play.py [hunk 3]
```diff
@@ -261,9 +256,9 @@ def play(
         else:
             action = key_code_to_action.get(tuple(sorted(game.pressed_keys)), noop)
             prev_obs = obs
-            obs, rew, done, info = env.step(action)
+            obs, rew, terminated, truncated, info = env.step(action)
```
Comment (pseudo-rnd-thoughts) [inline gym/utils/play.py:259]: need to add `done = termianted or truncated` after otherwise `done` is never updated

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Looks good, using control f "done" I found the following theses that need changing 

1. The README.md needs updating
2. passive environment checker.py line 229 updated to "rewrite the environment with (new) terminated / truncated step API."
3. play.py docstrings for `play()` and `PlayPlot`
4. vector_env.py docstring on line 141
5. docstring in autoreset.py line 10, 13 and 20
6. replace normalize.py
```
      if not self.is_vector_env:
          dones = terminateds or truncateds
      else:
          dones = np.bitwise_or(terminateds, truncateds)
```
with `dones = np.logical_or(terminateds, truncateds)` works with both bool and np.ndarray
7. Update wrappers step api compatbiility.py "Old step API refers to" to "(Old) Done step API refers to" and the same for "New step API" with "(New) Terminated / Truncated step API"

**Comment by pseudo-rnd-thoughts** [comment]
@RedTachyon This looks good to me, I did a document search for uses of "done" and "new_step_api", and all are correct now

File: gym/utils/step_api_compatibility.py [hunk 4]
```diff
@@ -128,33 +129,33 @@ def step_to_old_api(
 
 
 def step_api_compatibility(
-    step_returns: Union[NewStepType, OldStepType],
-    new_step_api: bool = False,
+    step_returns: Union[TerminatedTruncatedStepType, DoneStepType],
+    output_truncation_bool: bool = True,
```
Comment (RedTachyon) [inline gym/utils/step_api_compatibility.py:133]: I don't like this argument name, it's pretty unclear. I think there's a different name used earlier?

File: gym/wrappers/autoreset.py [hunk 5]
```diff
@@ -11,27 +10,27 @@ class AutoResetWrapper(gym.Wrapper):
     with new step API and ``(new_obs, final_reward, final_done, info)`` with the old step API.
```
Comment (RedTachyon) [inline gym/wrappers/autoreset.py:10]: I think the mention of the old style API should be removed?

File: gym/wrappers/normalize.py [hunk 6]
```diff
@@ -137,25 +127,16 @@ def __init__(
...
+        obs, rews, terminateds, truncateds, infos = self.env.step(action)
         if not self.is_vector_env:
             rews = np.array([rews])
         self.returns = self.returns * self.gamma + rews
         rews = self.normalize(rews)
-        if not self.is_vector_env:
-            dones = terminateds or truncateds
-        else:
-            dones = np.bitwise_or(terminateds, truncateds)
+        dones = np.logical_or(terminateds, truncateds)
```
Comment (RedTachyon) [inline gym/wrappers/normalize.py:135]: Why this change? It used to be either `or` or `np.bitwise_or`, now it's `np.logical_or`

File: gym/wrappers/step_api_compatibility.py [hunk 7]
```diff
@@ -11,37 +14,36 @@ class StepAPICompatibility(gym.Wrapper):
     New step API refers to step() method returning (observation, reward, terminated, truncated, info)
     (Refer to docs for details on the API change)
 
-    This wrapper is to be used to ease transition to new API and for backward compatibility.
-
     Args:
         env (gym.Env): the env to wrap. Can be in old or new API
-        new_step_api (bool): True to use env with new step API, False to use env with old step API. (False by default)
+        apply_step_compatibility (bool): Apply to convert environment to use new step API that returns two bools. (False by default)
```
Comment (RedTachyon) [inline gym/wrappers/step_api_compatibility.py:19]: Rephrase, the "Apply" sounds weird. Probably would be best as something like "Whether or not to [...]"

Also now I think it defaults to True

File: gym/wrappers/step_api_compatibility.py [hunk 8]
```diff
@@ -11,37 +14,36 @@ class StepAPICompatibility(gym.Wrapper):
...
+        self.output_truncation_bool = output_truncation_bool
+        if not self.output_truncation_bool:
             deprecation(
-                "Initializing environment in old step API which returns one bool instead of two. It is recommended to set `new_step_api=True` to use new step API. This will be the default behaviour in future."
+                "Initializing environment in old step API which returns one bool instead of two."
             )
 
     def step(self, action):
-        """Steps through the environment, returning 5 or 4 items depending on `new_step_api`.
+        """Steps through the environment, returning 5 or 4 items depending on `apply_step_compatibility`.
```
Comment (RedTachyon) [inline gym/wrappers/step_api_compatibility.py:46]: What I said in an earlier comment, sometimes it's `output_truncation_bool`, sometimes it's `apply_step_compatibility`, and it seems to have been mixed up here?

File: tests/vector/test_step_compatibility_vector.py [hunk 9]
```diff
@@ -1,88 +0,0 @@
...
-        obs = self.observation_space.sample()
-        rew = 0
-        terminated = False
-        truncated = False
-        info = {}
-        return obs, rew, terminated, truncated, info
-
-
-@pytest.mark.parametrize("VecEnv", [AsyncVectorEnv, SyncVectorEnv])
-def test_vector_step_compatibility_new_env(VecEnv):
```
Comment (RedTachyon) [inline tests/vector/test_step_compatibility_vector.py:43]: Do we not want to test this anymore?

**Review by RedTachyon — COMMENTED** [review]
Left a bunch of comments

File: gym/envs/registration.py [hunk 10]
```diff
@@ -684,26 +684,28 @@ def make(
...
     # Add the autoreset wrapper
     if autoreset:
-        env = AutoResetWrapper(env, new_step_api)
+        env = AutoResetWrapper(env)
 
     # Add human rendering wrapper
     if apply_human_rendering:
         env = HumanRendering(env)
 
+    # Add step API wrapper
```
[the remark this answers is not in the document]
Comment (RedTachyon) [reply gym/envs/registration.py:705]: (checked now, it doesn't work, at least assuming my understanding is correct)

File: gym/envs/registration.py [hunk 11]
```diff
@@ -557,7 +557,7 @@ def make(
         id: Name of the environment. Optionally, a module to import can be included, eg. 'module:Env-v0'
         max_episode_steps: Maximum length of an episode (TimeLimit wrapper).
         autoreset: Whether to automatically reset the environment after each episode (AutoResetWrapper).
-        new_step_api: Whether to use old or new step API (StepAPICompatibility wrapper). Will be removed at v1.0
+        apply_step_compatibility: Whether to use apply compatibility wrapper that converts step method to return two bools (StepAPICompatibility wrapper)
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gym/envs/registration.py:560]: Yes, I think we should keep a parameter in make to easily apply the compatibility wrapper

File: gym/envs/registration.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply gym/envs/registration.py:705]: Yes I think you are correct, this should occur after the environment checker in order of wrapper

File: gym/utils/step_api_compatibility.py (hunk 4 above)
Comment (pseudo-rnd-thoughts) [reply gym/utils/step_api_compatibility.py:133]: `to_termination_truncation_api` ? I agree it is not a great name but uncertain of a better one

File: gym/wrappers/normalize.py (hunk 6 above)
Comment (pseudo-rnd-thoughts) [reply gym/wrappers/normalize.py:135]: It simplifies the code. I checked, it works for both `True` and `np.array([True, False])`

File: gym/wrappers/step_api_compatibility.py (hunk 7 above)
Comment (pseudo-rnd-thoughts) [reply gym/wrappers/step_api_compatibility.py:19]: I think it should default to `False`. We **shouldn't** apply the compatibility wrapper by default

File: gym/wrappers/step_api_compatibility.py (hunk 8 above)
Comment (pseudo-rnd-thoughts) [reply gym/wrappers/step_api_compatibility.py:46]: `output_truncation_bool` is used in the step api compatibility wrapper while `apply_step_compatibility` is in `gym.make`

File: tests/vector/test_step_compatibility_vector.py (hunk 9 above)
Comment (pseudo-rnd-thoughts) [reply tests/vector/test_step_compatibility_vector.py:43]: This was purely for vector environments. These tests should be covered in the step api testing

File: gym/wrappers/step_api_compatibility.py (hunk 8 above)
Comment (pseudo-rnd-thoughts) [reply gym/wrappers/step_api_compatibility.py:46]: Thinking about it, it should be output_truncation_bool here

**Comment by pseudo-rnd-thoughts** [comment]
Due to @arjun-kg travelling, he is not able to fix these changes. Therefore, we will merge this PR with a follow up PR to make the relevant changes

**Comment by wookayin** [comment]
Having StepAPICompatibility sounds good for backward compatibility purposes despite the new default, but what about `env.reset()`?  I would suggest have a unified wrapper class for handling both `step` and `reset` API changes.
